crates/blockifier/src/execution/native/syscall_handler.rs

```diff
@@ -0,0 +1,318 @@
+    pub entry_point_selector: Felt,
+
+    // Execution results.
+    pub events: Vec<OrderedEvent>,
+    pub l2_to_l1_messages: Vec<OrderedL2ToL1Message>,
+    pub inner_calls: Vec<CallInfo>,
+
+    // Additional execution result info.
+    pub storage_read_values: Vec<Felt>,
+    pub accessed_storage_keys: HashSet<StorageKey, RandomState>,
+}
+
+impl<'state> NativeSyscallHandler<'state> {
+    pub fn new(
+        state: &'state mut dyn State,
+        caller_address: ContractAddress,
+        contract_address: ContractAddress,
+        entry_point_selector: EntryPointSelector,
+        execution_resources: &'state mut ExecutionResources,
+        execution_context: &'state mut EntryPointExecutionContext,
+    ) -> NativeSyscallHandler<'state> {
+        NativeSyscallHandler {
+            state,
+            caller_address,
+            contract_address,
+            entry_point_selector: entry_point_selector.0,
+            execution_resources,
+            execution_context,
+            events: Vec::new(),
+            l2_to_l1_messages: Vec::new(),
+            inner_calls: Vec::new(),
+            storage_read_values: Vec::new(),
+            accessed_storage_keys: HashSet::new(),
+        }
+    }
+
+    pub fn execute_inner_call(
+        &mut self,
+        entry_point: CallEntryPoint,
+        remaining_gas: &mut u128,
+    ) -> SyscallResult<CallInfo> {
+        let call_info = entry_point
+            .execute(self.state, self.execution_resources, self.execution_context)
+            .map_err(|e| encode_str_as_felts(&e.to_string()))?;
+        let retdata = call_info.execution.retdata.0.clone();
+
+        if call_info.execution.failed {
+            // In VM it's wrapped into `SyscallExecutionError::SyscallError`.
+            return Err(retdata);
+        }
+
+        self.update_remaining_gas(remaining_gas, &call_info);
+
+        self.inner_calls.push(call_info.clone());
+
+        Ok(call_info)
+    }
+
+    pub fn update_remaining_gas(&mut self, remaining_gas: &mut u128, call_info: &CallInfo) {
+        // Create a new variable with converted type.
+        let mut remaining_gas_u64 = u64::try_from(*remaining_gas).unwrap();
+
+        // Pass the reference to the function.
+        update_remaining_gas(&mut remaining_gas_u64, call_info);
+
+        // Change the remaining gas value.
+        *remaining_gas = u128::from(remaining_gas_u64);
+    }
+
+    // We need to have this function since in VM we have `execute_syscall` method, which is handling
+    // all gas-related logics in the native, syscalls are called directly, so we need to
+    // implement this logic here.
+    pub fn substract_syscall_gas_cost(
+        &mut self,
+        remaining_gas: &mut u128,
+        syscall_gas_cost: u64,
+    ) -> SyscallResult<()> {
+        // Refund `SYSCALL_BASE_GAS_COST` as it was pre-charged.
+        let required_gas =
+            u128::from(syscall_gas_cost - self.execution_context.gas_costs().syscall_base_gas_cost);
```

Comment:
Can `syscall_base_gas_cost` be bigger than `syscall_gas_cost`?

Reply:
Shouldn't be, this is how it's done in VM

Reply:
...and also in SiR